### DISCUSSION TITLE
Make MaltParameters animatable

### DISCUSSION THREAD

**Discussion by Kolupsy** [bnpr/Malt#336] [Feedback & Feature Requests]
I guess this is a deeper-reaching issue but I am not a fan of how malt parameters are not animatable. I think this may have to do with how they are stored. I guess right now they are custom properties. Is there a reason why they are not annotated properties from `bpy.props`? Because you can animate them and I feel like they are easier to manage.

**Comment by pragma37** [comment]
MaltProperties can be animated, but there's a limitation on the Blender with node properties. You can just override the property on the material side and animate that: https://github.com/bnpr/Malt/discussions/265#discussioncomment-2304666

> Is there a reason why they are not annotated properties from bpy.props? 

Then a new class for every new node would be needed, which would mean registering/unregistering nodes at runtime and crashing Blender.
It would also mean losing property overrides.

**Reply by Kolupsy** [reply]
> Then a new class for every new node would be needed

If you relocate all malt parameters to be stored on sockets instead of nodes, this would mean that you just need to dynamically add/remove sockets (which you are doing anyway) for the MaltFunctionNode for example (and I think all other nodes too). This would be more inline with blenders own design and most other custom node trees that I have seen.

I DID find a workaround though where I am able to use the depsgraph update system to drive malt properties which works fine too. Its a bit of a hassle to setup for a quick script but its acutally quite managable for an entire plugin.

This is how I implemented it: the whole thing only works because of the id property update function from the `rna_prop_ui` module:

https://github.com/Kolupsy/MaltShadingEssentials/blob/main/driver_registry.py

**Reply by pragma37** [reply]
> If you relocate all malt parameters to be stored on sockets instead of nodes, this would mean that you just need to dynamically add/remove sockets

While that's indeed a much better solution than registering classes at runtime, it would still mean duplicating all the MaltProperties functionality in a new system, which would be a lot of work just to avoid using material overrides.

**Reply by Kolupsy** [reply]
Hm I see. I thing material overrides dont always make sense from a design standpoint but since I just found a workaround, I personally dont care about that anymore
